packages/user/lib/user.ts

```diff
@@ -0,0 +1,143 @@
+  }
+
+  if (apiUser?.channels?.websocket) {
+    user.channels.websocket = apiUser.channels.websocket.map(
+      (apiSocket: WebSocketChannelResponse): WebsocketChannel => {
+        const socket = {
+          uri: apiSocket.uri,
+        } as WebsocketChannel;
+
+        if (apiSocket['content-type']) {
+          socket.contentType = apiSocket['content-type'];
+        }
+
+        if (apiSocket.headers) {
+          socket.headers = apiSocket.headers;
+        }
+
+        return socket;
+      },
+    );
+  }
+
+  return user;
+};
+
+const userToAPI = (user: UserType): UserResponse => {
+  const apiUser = Client.transformers.snakeCaseObjectKeys(user, true);
+  // preserve user properties
+  if (apiUser.properties?.custom_data) {
+    apiUser.properties.custom_data = user.properties.customData;
+  }
+
+  // Websockets will change when transformed
+  if (apiUser.channels?.websocket) {
+    apiUser.channels.websocket = user.channels.websocket.map(
+      (socket: WebsocketChannel): WebSocketChannelRequest => {
+        const apiSocket = {
+          uri: socket.uri,
+        } as WebSocketChannelRequest;
+
+        // restore content type
+        if (socket.contentType) {
+          apiSocket['content-type'] = socket.contentType;
+        }
+
+        // preserve headers
+        if (socket.headers) {
+          apiSocket.headers = socket.headers;
+        }
+
+        return apiSocket;
+      },
+    );
+  }
+
+  delete apiUser.id;
+
+  return apiUser;
+};
+
+export class User extends Client {
+  protected authType = AuthenticationType.JWT;
+
+  async *listAllUsers(
+    params: UserListParameters = {},
+  ): AsyncGenerator<UserType, void & UserType, undefined> {
+    let cursor = params.cursor;
+    do {
+      if (cursor) {
+        params.cursor = cursor;
+      }
+      const resp = await this.getUserPage(params);
+
+      yield* resp._embedded?.users.map(apiUserToUser);
+      const next = resp._links?.next?.href
+        ? new URL(resp._links.next.href)
+        : null;
+
+      cursor = next ? next.searchParams.get('cursor') : null;
+    } while (cursor);
+  }
+
+  async getUserPage(
+    params: UserListParameters = {},
+  ): Promise<UserPageResponse> {
+    const resp = await this.sendGetRequest<UserPageResponse>(
+      `${this.config.apiHost}/v1/users`,
+      Client.transformers.snakeCaseObjectKeys(params, true),
+    );
+
+    return resp.data;
+  }
+
+  async createUser(user: UserType): Promise<UserType> {
+    const resp = await this.sendPostRequest<UserResponse>(
+      `${this.config.apiHost}/v1/users`,
+      userToAPI(user),
+    );
+
+    return apiUserToUser(resp.data);
+  }
+
+  async getUser(userId: string): Promise<UserType> {
+    const resp = await this.sendGetRequest<UserResponse>(
+      `${this.config.apiHost}/v1/users/${userId}`,
+    );
+
+    return apiUserToUser(resp.data);
+  }
+
+  async updateUser(userId: string, user: UserType): Promise<UserType> {
```

Comment:
This should just take a UserType, and bail out if they don't pass an ID as part of the object. There's never a reason to update a user without an ID, and we don't want someone passing in an ID trying to update a user object with a different ID
```suggestion
  async updateUser(user: UserType): Promise<UserType> {
```

Reply:
Ive been wanting to have the methods go this way but didn't because that was how the other packages had the signature this way